sensing/pointcloud_preprocessor/src/downsample_filter/faster_voxel_grid_downsample_filter.cpp

```diff
@@ -0,0 +1,179 @@
+  if (
+    ((static_cast<std::int64_t>((max_point[0] - min_point[0]) * inverse_voxel_size_[0]) + 1) *
+     (static_cast<std::int64_t>((max_point[1] - min_point[1]) * inverse_voxel_size_[1]) + 1) *
+     (static_cast<std::int64_t>((max_point[2] - min_point[2]) * inverse_voxel_size_[2]) + 1)) >
+    static_cast<std::int64_t>(std::numeric_limits<std::int32_t>::max())) {
+    return false;
+  }
+
+  // Compute the minimum and maximum voxel coordinates
+  min_voxel[0] = static_cast<int>(std::floor(min_point[0] * inverse_voxel_size_[0]));
+  min_voxel[1] = static_cast<int>(std::floor(min_point[1] * inverse_voxel_size_[1]));
+  min_voxel[2] = static_cast<int>(std::floor(min_point[2] * inverse_voxel_size_[2]));
+  max_voxel[0] = static_cast<int>(std::floor(max_point[0] * inverse_voxel_size_[0]));
+  max_voxel[1] = static_cast<int>(std::floor(max_point[1] * inverse_voxel_size_[1]));
+  max_voxel[2] = static_cast<int>(std::floor(max_point[2] * inverse_voxel_size_[2]));
+
+  return true;
+}
+
+std::unordered_map<uint32_t, FasterVoxelGridDownsampleFilter::Centroid>
+FasterVoxelGridDownsampleFilter::calc_centroids_each_voxel(
+  const PointCloud2ConstPtr & input, const Eigen::Vector3f & max_voxel,
+  const Eigen::Vector3f & min_voxel)
+{
+  std::unordered_map<uint32_t, Centroid> voxel_centroid_map;
+  // Compute the number of divisions needed along all axis
+  Eigen::Vector3f div_b = max_voxel - min_voxel + Eigen::Vector3f::Ones();
+  // Set up the division multiplier
+  Eigen::Vector3i div_b_mul(1, div_b[0], div_b[0] * div_b[1]);
+
+  for (size_t global_offset = 0; global_offset + input->point_step <= input->data.size();
+       global_offset += input->point_step) {
+    Eigen::Vector3f point = get_point_from_global_offset(input, global_offset);
+    if (std::isfinite(point[0]) && std::isfinite(point[1]), std::isfinite(point[2])) {
+      // Calculate the voxel index to which the point belongs
+      int ijk0 = static_cast<int>(std::floor(point[0] * inverse_voxel_size_[0]) - min_voxel[0]);
+      int ijk1 = static_cast<int>(std::floor(point[1] * inverse_voxel_size_[1]) - min_voxel[1]);
+      int ijk2 = static_cast<int>(std::floor(point[2] * inverse_voxel_size_[2]) - min_voxel[2]);
+      uint32_t voxel_id = ijk0 * div_b_mul[0] + ijk1 * div_b_mul[1] + ijk2 * div_b_mul[2];
+
+      // Add the point to the corresponding centroid
+      if (voxel_centroid_map.find(voxel_id) == voxel_centroid_map.end()) {
+        voxel_centroid_map[voxel_id] = Centroid(point[0], point[1], point[2]);
+      } else {
+        voxel_centroid_map[voxel_id].add_point(point[0], point[1], point[2]);
+      }
+    }
+  }
+
+  return voxel_centroid_map;
+}
+
+void FasterVoxelGridDownsampleFilter::copy_centroids_to_output(
+  std::unordered_map<uint32_t, Centroid> & voxel_centroid_map, PointCloud2 & output,
+  const TransformInfo & transform_info)
+{
+  size_t output_data_size = 0;
+  for (auto & pair : voxel_centroid_map) {
```

Comment:
```suggestion
  for (const auto & pair : voxel_centroid_map) {
```
Is it better to use const?

Reply:
Addressed in https://github.com/autowarefoundation/autoware.universe/pull/3679/commits/29c88b2b8e36c5ccb8ac540e0a3f8d0270b4a867.